File: src/lib/utils/connection.ts
import type { ConnectionInput, DatabaseType } from '$lib/rpc';
import { dialectCapabilities } from '$lib/utils/dialect';

export function generateConnectionString(form: ConnectionInput): string {
	try {
		if (form.databaseType === 'sqlite') {
			return form.database ? `sqlite://${form.database}` : '';
		}
		const protocol =
			form.databaseType === 'mysql'
				? 'mysql'
				: form.databaseType === 'mssql'
					? 'sqlserver'
					: 'postgres';
		const userPart = form.user ? encodeURIComponent(form.user) : '';
		const passPart = form.password ? `:${encodeURIComponent(form.password)}` : '';
		const authPart = userPart ? `${userPart}${passPart}@` : '';
		const hostPart = form.host || 'localhost';
		const portPart = form.port ? `:${form.port}` : '';
		const dbPart = form.database ? `/${form.database}` : '';
		return `${protocol}://${authPart}${hostPart}${portPart}${dbPart}`;
	} catch {
		return '';
	}
}

export function connectionStringPlaceholder(databaseType: DatabaseType): string {
	if (databaseType === 'sqlite') return 'sqlite://C:/path/to/db.sqlite';
	if (databaseType === 'mysql') return 'mysql://root:password@localhost:3306/mydb';
	if (databaseType === 'mssql') return 'sqlserver://sa:password@localhost:1433/master';
	return 'postgres://postgres:password@localhost:5432/postgres';
}

export function parseConnectionString(
	value: string,
	current: ConnectionInput,
): Partial<ConnectionInput> | null {
	const val = value.trim();
	if (!val) return null;
	try {
		if (val.startsWith('sqlite://')) {
			return { databaseType: 'sqlite', database: val.slice(9) };
		}
		if (/^(jdbc:)?sqlserver:/i.test(val) || val.toLowerCase().startsWith('mssql:')) {
			return parseMssqlConnectionString(val, current);
		}
		if (looksLikeAdoNet(val)) {
			return parseAdoNetConnectionString(val, current);
		}
		const url = new URL(val);
		let databaseType = current.databaseType;
		if (url.protocol.includes('postgres')) databaseType = 'postgres';
		else if (url.protocol.includes('mysql')) databaseType = 'mysql';
		else if (url.protocol.includes('mssql') || url.protocol.includes('sqlserver')) {
			databaseType = 'mssql';
		}
		return {
			databaseType,
			host: url.hostname || current.host,
			port: url.port
				? Number(url.port)
				: dialectCapabilities(databaseType).defaultPort,
			user: url.username ? decodeURIComponent(url.username) : current.user,
			password: url.password ? decodeURIComponent(url.password) : current.password,
			database:
				url.pathname && url.pathname.length > 1
					? url.pathname.slice(1)
					: current.database,
		};
	} catch {
		return null;
	}
}

export function defaultsForType(databaseType: DatabaseType): ConnectionInput {
	return normalizeConnectionInput({ databaseType });
}

export const DATABASE_ENGINES: Array<{
	value: DatabaseType;
	label: string;
	detail: string;
}> = [
	{ value: 'postgres', label: 'PostgreSQL', detail: 'Relational' },
	{ value: 'mysql', label: 'MySQL', detail: 'Relational' },
	{ value: 'mssql', label: 'SQL Server', detail: 'Relational' },
	{ value: 'sqlite', label: 'SQLite', detail: 'Local file' },
];

export const ENGINE_KEY: Record<DatabaseType, string> = {
	sqlite: '#0f80cc',
	postgres: '#336791',
	mssql: '#cc2927',
	mysql: '#00758f',
};

function looksLikeAdoNet(value: string): boolean {
	return /(?:^|;)\s*(server|data source|initial catalog|database|user id|uid|password|pwd|encrypt|trustservercertificate)\s*=/i.test(
		value,
	);
}

function adoNetMap(value: string): Map<string, string> {
	const map = new Map<string, string>();
	for (const part of value.split(';')) {
		const trimmed = part.trim();
		if (!trimmed) continue;
		const eq = trimmed.indexOf('=');
		if (eq <= 0) continue;
		map.set(trimmed.slice(0, eq).trim().toLowerCase(), trimmed.slice(eq + 1).trim());
	}
	return map;
}

function parseAdoNetConnectionString(
	value: string,
	current: ConnectionInput,
): Partial<ConnectionInput> {
	const map = adoNetMap(value);
	const server = map.get('server') || map.get('data source') || '';
	let host = current.host;
	let port = dialectCapabilities('mssql').defaultPort;
	if (server) {
		const cleaned = server.replace(/^tcp:/i, '');
		const comma = cleaned.lastIndexOf(',');
		if (comma >= 0) {
			host = cleaned.slice(0, comma).trim() || host;
			const parsedPort = Number(cleaned.slice(comma + 1).trim());
			if (Number.isFinite(parsedPort) && parsedPort > 0) port = parsedPort;
		} else {
			host = cleaned.trim() || host;
		}
	}
	const encrypt = map.get('encrypt');
	const trust = map.get('trustservercertificate');
	return {
		databaseType: 'mssql',
		host,
		port,
		user: map.get('user id') || map.get('uid') || current.user,
		password: map.get('password') || map.get('pwd') || current.password,
		database: map.get('database') || map.get('initial catalog') || current.database,
		ssl: encrypt ? !/^(no|false|off|disable|optional)$/i.test(encrypt) : true,
		sslInsecure: trust ? /^(yes|true|1)$/i.test(trust) : current.sslInsecure,
		useConnectionString: true,
		connectionString: value,
	};
}

function parseMssqlConnectionString(
	value: string,
	current: ConnectionInput,
): Partial<ConnectionInput> | null {
	const jdbc = value.match(/^jdbc:sqlserver:\/\/([^;]+)(?:;(.*))?$/i);
	if (jdbc) {
		const hostPort = jdbc[1] ?? '';
		const rest = jdbc[2] ? `Server=${hostPort};${jdbc[2]}` : `Server=${hostPort}`;
		return parseAdoNetConnectionString(rest, current);
	}
	try {
		const normalized = value.replace(/^mssql:/i, 'sqlserver:');
		const url = new URL(normalized);
		if (!url.protocol.includes('sqlserver') && !url.protocol.includes('mssql')) return null;
		return {
			databaseType: 'mssql',
			host: url.hostname || current.host,
			port: url.port ? Number(url.port) : dialectCapabilities('mssql').defaultPort,
			user: url.username ? decodeURIComponent(url.username) : current.user,
			password: url.password ? decodeURIComponent(url.password) : current.password,
			database:
				url.pathname && url.pathname.length > 1
					? url.pathname.slice(1)
					: current.database,
		};
	} catch {
		return parseAdoNetConnectionString(value, current);
	}
}

export function withDatabaseType(
	current: ConnectionInput,
	nextType: DatabaseType,
): ConnectionInput {
	if (current.databaseType === nextType) return current;
	const defaults = defaultsForType(nextType);
	const previousDefaultName = defaultsForType(current.databaseType).name;
	const keepName = Boolean(current.name.trim()) && current.name !== previousDefaultName;
	return {
		...defaults,
		name: keepName ? current.name : defaults.name,
		password: nextType === 'sqlite' ? '' : current.password,
		ssl: nextType === 'sqlite' ? false : nextType === 'mssql' ? true : current.ssl,
		sslInsecure:
			nextType === 'sqlite'
				? false
				: nextType === 'mssql'
					? true
					: current.ssl && Boolean(current.sslInsecure),
	};
}

export function withSqliteFile(current: ConnectionInput, path: string): ConnectionInput {
	const fileName = path.replaceAll('\\', '/').split('/').pop() ?? '';
	const defaultName = defaultsForType('sqlite').name;
	const nextName =
		!current.name.trim() || current.name === defaultName
			? fileName.replace(/\.(sqlite|sqlite3|db)$/i, '') || current.name
			: current.name;
	return { ...current, databaseType: 'sqlite', database: path, name: nextName };
}

export function withParsedConnectionString(
	current: ConnectionInput,
	value: string,
): ConnectionInput | null {
	const parsed = parseConnectionString(value, current);
	if (!parsed) return null;
	return normalizeConnectionInput({ ...current, ...parsed });
}

export function connectionSubtitle(connection: ConnectionInput): string {
	if (connection.databaseType === 'sqlite') {
		return connection.database || 'local file';
	}
	const host = connection.host || 'localhost';
	const port = connection.port ? `:${connection.port}` : '';
	const db = connection.database ? ` / ${connection.database}` : '';
	return `${host}${port}${db}`;
}

export function connectionEngineLabel(databaseType: DatabaseType): string {
	if (databaseType === 'mysql') return 'mysql';
	if (databaseType === 'sqlite') return 'sqlite';
	if (databaseType === 'mssql') return 'mssql';
	return 'postgres';
}

export function connectionMetaLine(connection: ConnectionInput): string {
	const engine = connectionEngineLabel(connection.databaseType);
	if (connection.databaseType === 'sqlite') {
		const file = connection.database.split(/[/\\]/).pop() || 'local file';
		return `${engine} · ${file}`;
	}
	return `${engine} · ${connection.host || 'localhost'}`;
}

export const RECENT_CONNECTIONS_KEY = 'querycastle.recentConnections.v1';

export function loadRecentConnectionNames(): string[] {
	if (typeof localStorage === 'undefined') return [];
	try {
		const raw = localStorage.getItem(RECENT_CONNECTIONS_KEY);
		if (!raw) return [];
		const parsed = JSON.parse(raw) as unknown;
		if (!Array.isArray(parsed)) return [];
		return parsed.filter((item): item is string => typeof item === 'string' && item.length > 0);
	} catch {
		return [];
	}
}

export function rememberRecentConnection(name: string): string[] {
	const trimmed = name.trim();
	if (!trimmed) return loadRecentConnectionNames();
	const next = [trimmed, ...loadRecentConnectionNames().filter((item) => item !== trimmed)].slice(0, 8);
	try {
		localStorage.setItem(RECENT_CONNECTIONS_KEY, JSON.stringify(next));
	} catch {
		// ignore quota / private-mode failures
	}
	return next;
}

export function normalizeConnectionInput(
	input: Partial<ConnectionInput>,
): ConnectionInput {
	const databaseType: DatabaseType =
		input.databaseType === 'mysql' ||
		input.databaseType === 'sqlite' ||
		input.databaseType === 'mssql'
			? input.databaseType
			: 'postgres';
	const defaults = dialectCapabilities(databaseType);
	const sslDefault = databaseType === 'mssql' ? true : false;
	const sslInsecureDefault = databaseType === 'mssql' ? true : false;

	return {
		databaseType,
		name: input.name ?? defaults.defaultName,
		host: input.host ?? (databaseType === 'sqlite' ? '' : 'localhost'),
		port: input.port ?? defaults.defaultPort,
		user: input.user ?? defaults.defaultUser,
		password: input.password ?? '',
		database: input.database ?? defaults.defaultDatabase,
		ssl: databaseType === 'sqlite' ? false : (input.ssl ?? sslDefault),
		sslInsecure:
			databaseType === 'sqlite' ? false : (input.sslInsecure ?? sslInsecureDefault),
		useConnectionString: input.useConnectionString ?? false,
		connectionString: input.connectionString ?? '',
	};
}

function setConnectionStringPassword(raw: string, password: string): string {
	try {
		const url = new URL(raw);
		url.password = password;
		return url.toString();
	} catch {
		return raw;
	}
}

function stripConnectionStringPassword(raw: string): string {
	try {
		const url = new URL(raw);
		if (!url.password) return raw;
		url.password = '';
		return url.toString();
	} catch {
		return raw;
	}
}

export function passwordFromConnection(connection: ConnectionInput): string {
	if (connection.password) return connection.password;
	const raw = connection.connectionString?.trim() ?? '';
	if (!raw) return '';
	try {
		return new URL(raw).password || '';
	} catch {
		return '';
	}
}

export function stripConnectionSecrets(connection: ConnectionInput): ConnectionInput {
	return {
		...connection,
		password: '',
		connectionString: connection.connectionString
			? stripConnectionStringPassword(connection.connectionString)
			: connection.connectionString,
	};
}

export function injectConnectionPassword(
	connection: ConnectionInput,
	password: string,
): ConnectionInput {
	const next: ConnectionInput = { ...connection, password };
	if (next.useConnectionString && next.connectionString?.trim()) {
		next.connectionString = setConnectionStringPassword(next.connectionString, password);
	}
	return next;
}

export async function migrateSavedConnectionSecrets(params: {
	connections: ConnectionInput[];
	secretSet: (name: string, password: string) => Promise<void>;
}): Promise<{ connections: ConnectionInput[]; changed: boolean }> {
	let changed = false;
	const next: ConnectionInput[] = [];
	for (const connection of params.connections) {
		const password = passwordFromConnection(connection);
		const stripped = stripConnectionSecrets(connection);
		const needsStrip =
			stripped.password !== connection.password ||
			stripped.connectionString !== connection.connectionString;
		if (password && connection.name.trim()) {
			try {
				await params.secretSet(connection.name, password);
				next.push(stripped);
				changed = true;
			} catch {
				next.push(connection);
			}
		} else if (needsStrip) {
			next.push(stripped);
			changed = true;
		} else {
			next.push(stripped);
		}
	}
	return { connections: next, changed };
}

File: src-tauri/src/core/connection.rs
use url::Url;

use crate::core::error::DbError;
use crate::core::types::{ConnectionInput, DatabaseType};

fn default_port(database_type: DatabaseType) -> u16 {
    match database_type {
        DatabaseType::Postgres => 5432,
        DatabaseType::Mysql => 3306,
        DatabaseType::Sqlite => 0,
        DatabaseType::Mssql => 1433,
    }
}

fn default_user(database_type: DatabaseType) -> &'static str {
    match database_type {
        DatabaseType::Postgres => "postgres",
        DatabaseType::Mysql => "root",
        DatabaseType::Sqlite => "",
        DatabaseType::Mssql => "sa",
    }
}

fn default_database(database_type: DatabaseType) -> &'static str {
    match database_type {
        DatabaseType::Postgres => "postgres",
        DatabaseType::Mysql => "mysql",
        DatabaseType::Sqlite => "main",
        DatabaseType::Mssql => "master",
    }
}

fn overlay_session_flags(
    mut built: ConnectionInput,
    source: &ConnectionInput,
) -> Result<ConnectionInput, DbError> {
    built.read_only = source.read_only;
    if built.database_type == DatabaseType::Sqlite {
        built.ssh_enabled = false;
        built.ssh_host = String::new();
        built.ssh_port = 22;
        built.ssh_user = String::new();
        built.ssh_password = String::new();
        built.ssh_private_key_path = String::new();
        built.ssh_key_passphrase = String::new();
        return Ok(built);
    }
    built.ssh_enabled = source.ssh_enabled;
    built.ssh_host = source.ssh_host.trim().to_string();
    built.ssh_port = if source.ssh_port == 0 { 22 } else { source.ssh_port };
    built.ssh_user = source.ssh_user.trim().to_string();
    built.ssh_password = source.ssh_password.clone();
    built.ssh_private_key_path = source.ssh_private_key_path.trim().to_string();
    built.ssh_key_passphrase = source.ssh_key_passphrase.clone();
    if built.ssh_enabled {
        if built.ssh_host.is_empty() || built.ssh_user.is_empty() {
            return Err(DbError::validation("SSH host and user are required"));
        }
        if built.ssh_password.is_empty() && built.ssh_private_key_path.is_empty() {
            return Err(DbError::validation(
                "SSH password or private key is required",
            ));
        }
    }
    Ok(built)
}

pub(crate) fn rewrite_for_local_tunnel(input: &ConnectionInput, local_port: u16) -> ConnectionInput {
    let mut next = input.clone();
    next.host = "127.0.0.1".to_string();
    next.port = local_port;
    if next.use_connection_string && !next.connection_string.trim().is_empty() {
        next.connection_string = rewrite_endpoint(&next.connection_string, local_port);
    }
    next
}

fn rewrite_endpoint(raw: &str, local_port: u16) -> String {
    let trimmed = raw.trim();
    if let Some(rest) = trimmed
        .strip_prefix("jdbc:sqlserver://")
        .or_else(|| trimmed.strip_prefix("jdbc:SQLServer://"))
    {
        let (hostport, params) = rest.split_once(';').unwrap_or((rest, ""));
        let _ = hostport;
        return if params.is_empty() {
            format!("jdbc:sqlserver://127.0.0.1:{local_port}")
        } else {
            format!("jdbc:sqlserver://127.0.0.1:{local_port};{params}")
        };
    }
    if let Ok(mut url) = Url::parse(trimmed) {
        if url.host_str().is_some() {
            let _ = url.set_host(Some("127.0.0.1"));
            let _ = url.set_port(Some(local_port));
            return url.to_string();
        }
    }
    let mut parts: Vec<String> = Vec::new();
    let mut replaced = false;
    for part in trimmed.split(';') {
        let item = part.trim();
        if item.is_empty() {
            continue;
        }
        let Some((key, _)) = item.split_once('=') else {
            parts.push(item.to_string());
            continue;
        };
        if key.eq_ignore_ascii_case("server") || key.eq_ignore_ascii_case("data source") {
            parts.push(format!("{}=127.0.0.1,{local_port}", key.trim()));
            replaced = true;
        } else {
            parts.push(item.to_string());
        }
    }
    if replaced {
        parts.join(";")
    } else {
        trimmed.to_string()
    }
}

pub(crate) fn normalize_connection_input(input: ConnectionInput) -> Result<ConnectionInput, DbError> {
    let source = input.clone();
    let database_type = input.database_type;

    if input.use_connection_string {
        let raw = input.connection_string.trim().to_string();
        if raw.is_empty() {
            return Err(DbError::validation("Connection string is required"));
        }

        if database_type == DatabaseType::Mssql {
            let lower = raw.to_ascii_lowercase();
            if lower.starts_with("jdbc:sqlserver:") {
                return normalize_mssql_jdbc_string(input, raw);
            }
            if !lower.starts_with("sqlserver://") && !lower.starts_with("mssql://") {
                return normalize_mssql_ado_string(input, raw);
            }
        }

        let parsed = Url::parse(&raw).map_err(|_| DbError::validation("Invalid connection string format"))?;
        let valid_scheme = match database_type {
            DatabaseType::Postgres => parsed.scheme() == "postgresql" || parsed.scheme() == "postgres",
            DatabaseType::Mysql => parsed.scheme() == "mysql",
            DatabaseType::Sqlite => parsed.scheme() == "sqlite",
            DatabaseType::Mssql => parsed.scheme() == "sqlserver" || parsed.scheme() == "mssql",
        };
        if !valid_scheme {
            return Err(DbError::validation(match database_type {
                DatabaseType::Postgres => "Connection string must start with postgresql:// or postgres://",
                DatabaseType::Mysql => "Connection string must start with mysql://",
                DatabaseType::Sqlite => "Connection string must start with sqlite://",
                DatabaseType::Mssql => {
                    "Connection string must be sqlserver://, mssql://, or an ADO.NET string"
                }
            }));
        }

        let database = match database_type {
            DatabaseType::Sqlite => {
                let path = parsed.path().trim_start_matches('/').to_string();
                if path.is_empty() {
                    return Err(DbError::validation("Database path is missing in connection string"));
                }
                path
            }
            _ => {
                let db = parsed.path().trim_start_matches('/').to_string();
                if db.is_empty() {
                    default_database(database_type).to_string()
                } else {
                    db
                }
            }
        };

        let resolved_port = parsed.port().unwrap_or(default_port(database_type));
        let ssl_from_url = parsed.query_pairs().find_map(|(key, val)| {
            if key == "sslmode" {
                let mode = val.to_lowercase();
                Some(["require", "verify-ca", "verify-full"].contains(&mode.as_str()))
            } else {
                None
            }
        });

        return overlay_session_flags(ConnectionInput {
            database_type,
            name: if input.name.trim().is_empty() {
                database.clone()
            } else {
                input.name.trim().to_string()
            },
            host: if parsed.host_str().unwrap_or_default().is_empty() {
                if database_type == DatabaseType::Sqlite {
                    String::new()
                } else {
                    "localhost".to_string()
                }
            } else {
                parsed.host_str().unwrap_or_default().to_string()
            },
            port: resolved_port,
            user: if parsed.username().is_empty() {
                default_user(database_type).to_string()
            } else {
                parsed.username().to_string()
            },
            password: parsed.password().unwrap_or_default().to_string(),
            database,
            ssl: if database_type == DatabaseType::Sqlite {
                false
            } else {
                ssl_from_url.unwrap_or(input.ssl)
            },
            ssl_insecure: if database_type == DatabaseType::Sqlite {
                false
            } else {
                input.ssl_insecure
            },
            use_connection_string: true,
            connection_string: raw,
            ..ConnectionInput::default()
        }, &source);
    }

    if database_type == DatabaseType::Sqlite {
        if input.database.trim().is_empty() {
            return Err(DbError::validation("Database path is required for SQLite"));
        }
    } else if input.host.trim().is_empty() || input.user.trim().is_empty() {
        return Err(DbError::validation("Host and user are required"));
    }

    overlay_session_flags(ConnectionInput {
        database_type,
        name: if input.name.trim().is_empty() {
            if !input.database.trim().is_empty() {
                input.database.clone()
            } else if !input.host.trim().is_empty() {
                input.host.trim().to_string()
            } else {
                default_database(database_type).to_string()
            }
        } else {
            input.name.trim().to_string()
        },
        host: if database_type == DatabaseType::Sqlite {
            String::new()
        } else {
            input.host
        },
        port: if input.port == 0 {
            default_port(database_type)
        } else {
            input.port
        },
        user: if input.user.trim().is_empty() {
            default_user(database_type).to_string()
        } else {
            input.user
        },
        password: input.password,
        database: if input.database.trim().is_empty() {
            default_database(database_type).to_string()
        } else {
            input.database
        },
        ssl: if database_type == DatabaseType::Sqlite { false } else { input.ssl },
        ssl_insecure: if database_type == DatabaseType::Sqlite {
            false
        } else {
            input.ssl_insecure
        },
        use_connection_string: false,
        connection_string: String::new(),
        ..ConnectionInput::default()
    }, &source)
}

pub(crate) fn with_new_database(connection: &ConnectionInput, new_database: &str) -> ConnectionInput {
    let new_db = new_database.trim().to_string();
    if connection.use_connection_string {
        let raw = connection.connection_string.trim();
        if !raw.is_empty() {
            if let Ok(mut url) = Url::parse(raw) {
                match connection.database_type {
                    DatabaseType::Postgres | DatabaseType::Mysql | DatabaseType::Mssql => {
                        url.set_path(&format!("/{}", new_db));
                        return ConnectionInput {
                            database: new_db.clone(),
                            connection_string: url.to_string(),
                            ..connection.clone()
                        };
                    }
                    DatabaseType::Sqlite => {
                        url.set_path(&format!("/{}", new_db.trim_start_matches('/')));
                        return ConnectionInput {
                            database: new_db.clone(),
                            connection_string: url.to_string(),
                            ..connection.clone()
                        };
                    }
                }
            }
        }
    }
    ConnectionInput {
        database: new_db,
        ..connection.clone()
    }
}

fn normalize_mssql_jdbc_string(
    input: ConnectionInput,
    raw: String,
) -> Result<ConnectionInput, DbError> {
    let rest = raw
        .split_once("://")
        .map(|(_, tail)| tail)
        .unwrap_or(raw.as_str());
    let (hostport, params) = rest.split_once(';').unwrap_or((rest, ""));
    let ado = if hostport.trim().is_empty() {
        params.to_string()
    } else {
        format!("Server={hostport};{params}")
    };
    normalize_mssql_ado_string(input, ado).map(|mut parsed| {
        parsed.connection_string = raw;
        parsed
    })
}

fn normalize_mssql_ado_string(
    input: ConnectionInput,
    raw: String,
) -> Result<ConnectionInput, DbError> {
    let source = input.clone();
    let mut host = input.host;
    let mut port = if input.port == 0 {
        default_port(DatabaseType::Mssql)
    } else {
        input.port
    };
    let mut user = input.user;
    let mut password = input.password;
    let mut database = input.database;
    let mut ssl = input.ssl;
    let mut ssl_insecure = input.ssl_insecure;

    for part in raw.split(';') {
        let trimmed = part.trim();
        if trimmed.is_empty() {
            continue;
        }
        let Some((key, value)) = trimmed.split_once('=') else {
            continue;
        };
        let key = key.trim().to_ascii_lowercase();
        let value = value.trim();
        match key.as_str() {
            "server" | "data source" => {
                let cleaned = value.trim_start_matches("tcp:");
                if let Some((h, p)) = cleaned.rsplit_once(',') {
                    host = h.trim().to_string();
                    if let Ok(parsed) = p.trim().parse::<u16>() {
                        port = parsed;
                    }
                } else {
                    host = cleaned.trim().to_string();
                }
            }
            "database" | "initial catalog" | "databasename" => database = value.to_string(),
            "user id" | "uid" => user = value.to_string(),
            "password" | "pwd" => password = value.to_string(),
            "encrypt" => {
                ssl = !matches!(
                    value.to_ascii_lowercase().as_str(),
                    "no" | "false" | "off" | "disable" | "optional"
                );
            }
            "trustservercertificate" => {
                ssl_insecure = matches!(value.to_ascii_lowercase().as_str(), "yes" | "true" | "1");
            }
            _ => {}
        }
    }

    if database.trim().is_empty() {
        database = default_database(DatabaseType::Mssql).to_string();
    }
    if user.trim().is_empty() {
        user = default_user(DatabaseType::Mssql).to_string();
    }
    if host.trim().is_empty() {
        host = "localhost".to_string();
    }

    overlay_session_flags(ConnectionInput {
        database_type: DatabaseType::Mssql,
        name: if input.name.trim().is_empty() {
            database.clone()
        } else {
            input.name.trim().to_string()
        },
        host,
        port,
        user,
        password,
        database,
        ssl,
        ssl_insecure,
        use_connection_string: true,
        connection_string: raw,
        ..ConnectionInput::default()
    }, &source)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample() -> ConnectionInput {
        ConnectionInput {
            database_type: DatabaseType::Postgres,
            name: "n".into(),
            host: "localhost".into(),
            port: 5432,
            user: "postgres".into(),
            password: "secret".into(),
            database: "postgres".into(),
            ssl: true,
            ssl_insecure: true,
            use_connection_string: false,
            connection_string: String::new(),
            ..ConnectionInput::default()
        }
    }

    #[test]
    fn preserves_ssl_insecure() {
        let out = normalize_connection_input(sample()).unwrap();
        assert!(out.ssl);
        assert!(out.ssl_insecure);
    }

    #[test]
    fn sqlite_forces_ssl_off() {
        let mut input = sample();
        input.database_type = DatabaseType::Sqlite;
        input.database = "C:/tmp/x.db".into();
        let out = normalize_connection_input(input).unwrap();
        assert!(!out.ssl);
        assert!(!out.ssl_insecure);
    }

    #[test]
    fn mssql_url_and_ado_strings() {
        let mut input = sample();
        input.database_type = DatabaseType::Mssql;
        input.use_connection_string = true;
        input.connection_string = "sqlserver://sa:pw@db.example:14333/app".into();
        let out = normalize_connection_input(input.clone()).unwrap();
        assert_eq!(out.host, "db.example");
        assert_eq!(out.port, 14333);
        assert_eq!(out.user, "sa");
        assert_eq!(out.password, "pw");
        assert_eq!(out.database, "app");

        input.connection_string =
            "Server=localhost,1433;Database=shop;User Id=sa;Password=secret;Encrypt=yes;TrustServerCertificate=yes"
                .into();
        let ado = normalize_connection_input(input).unwrap();
        assert_eq!(ado.host, "localhost");
        assert_eq!(ado.port, 1433);
        assert_eq!(ado.database, "shop");
        assert!(ado.ssl);
        assert!(ado.ssl_insecure);
    }

    #[test]
    fn preserves_read_only_and_rewrites_tunnel_endpoint() {
        let mut input = sample();
        input.read_only = true;
        input.ssh_enabled = true;
        input.ssh_host = "bastion".into();
        input.ssh_user = "ubuntu".into();
        input.ssh_password = "pw".into();
        let out = normalize_connection_input(input).unwrap();
        assert!(out.read_only);
        assert!(out.ssh_enabled);
        assert_eq!(out.ssh_port, 22);

        let tunneled = rewrite_for_local_tunnel(&out, 2345);
        assert_eq!(tunneled.host, "127.0.0.1");
        assert_eq!(tunneled.port, 2345);
        assert_eq!(tunneled.ssh_host, "bastion");
    }
}
